samza-core/src/main/java/org/apache/samza/operators/impl/OperatorImplGraph.java

```diff
@@ -175,8 +176,15 @@ OperatorImpl createAndRegisterOperatorImpl(OperatorSpec prevOperatorSpec, Operat
       return operatorImpl;
     } else {
       // the implementation corresponding to operatorSpec has already been instantiated
-      // and registered, so we do not need to traverse the DAG further.
-      return operatorImpls.get(operatorSpec.getOpId());
+      // and registered. We still need to traverse the DAG further to register the input streams.
+      OperatorImpl operatorImpl = operatorImpls.get(operatorSpec.getOpId());
+      operatorImpl.registerInputStream(inputStream);
+
+      Collection<OperatorSpec> registeredSpecs = operatorSpec.getRegisteredOperatorSpecs();
+      registeredSpecs.forEach(registeredSpec -> {
+          createAndRegisterOperatorImpl(operatorSpec, registeredSpec, inputStream, config, context);
+        });
```

Comment:
wondering if this logic is needed here? Is the following equivalent?

I'd flip the statement that adds it to the map, and add it to the `operatorImpls ` after the recursive call returns?

```
  OperatorImpl createAndRegisterOperatorImpl(OperatorSpec prevOperatorSpec, OperatorSpec operatorSpec,
      SystemStream inputStream, Config config, TaskContext context) {
    if (!operatorImpls.containsKey(operatorSpec.getOpId()) || operatorSpec instanceof JoinOperatorSpec) {
      OperatorImpl operatorImpl = createOperatorImpl(prevOperatorSpec, operatorSpec, config, context);
      operatorImpl.init(config, context);
      operatorImpl.registerInputStream(inputStream);

      Collection<OperatorSpec> registeredSpecs = operatorSpec.getRegisteredOperatorSpecs();
      registeredSpecs.forEach(registeredSpec -> {
          OperatorImpl nextImpl = createAndRegisterOperatorImpl(operatorSpec, registeredSpec, inputStream, config, context);
          operatorImpl.registerNextOperator(nextImpl);
        });
      operatorImpls.put(operatorImpl.getOpImplId(), operatorImpl);
      return operatorImpl;
    } else {
      return operatorImpls.get(operatorSpec.getOpId());
    }
  }
```

Reply:
This logic is needed since I need to further traverse the graph even if the operator is already registered. That way the input stream can be registered to the downstream operators.